### PR TITLE
Create token implementation

### PR DESCRIPTION
**Pull request by alistair-singh**
Cumulus Companion: https://github.com/Snowfork/cumulus/pull/29

Resolves: SNO-490, SNO-491, SNO-431, SNO-450

### CODE REVIEW

File: parachain/primitives/router/src/lib.rs [hunk 1]
```diff
@@ -37,6 +43,25 @@ pub trait ConvertMessage {
...
+#[derive(Clone, Eq, PartialEq, Encode)]
+pub enum StatemineCall {
+	#[codec(index = 53u8)]
+	Assets(AssetsCall),
+}
+
+pub type Signature = MultiSignature;
+type AccountId = <<Signature as Verify>::Signer as IdentifyAccount>::AccountId;
+type AccountIndex = ();
+type Address = MultiAddress<AccountId, AccountIndex>;
```
Comment (vgeddes) [inline parachain/primitives/router/src/lib.rs:55]: None of this is necessary. An AccountId is usually configured to be 32 byte array: `[u8; 32]`.

File: parachain/primitives/router/src/lib.rs [hunk 2]
```diff
@@ -37,6 +43,25 @@ pub trait ConvertMessage {
...
+
+#[derive(Clone, Eq, PartialEq, Encode)]
+pub enum AssetsCall {
+	#[codec(index = 0u8)]
+	Create { asset_id: MultiLocation, owner: Address, min_balance: u128 },
```
Comment (vgeddes) [inline parachain/primitives/router/src/lib.rs:60]:
```suggestion
	Create { asset_id: MultiLocation, owner: [u8; 32], min_balance: u128 },
```

File: parachain/primitives/router/src/lib.rs [hunk 3]
```diff
@@ -62,7 +87,50 @@ where
...
+					fun: Fungible(buy_execution_fee_amount),
+				};
+
+				let instructions: Vec<Instruction<()>> = vec![
+					UniversalOrigin(GlobalConsensus(network)),
+					DescendOrigin(X1(Junction::AccountKey20 { network: None, key: origin.into() })),
+					WithdrawAsset(buy_execution_fee.clone().into()),
+					BuyExecution { fees: buy_execution_fee.clone(), weight_limit: Unlimited },
+					Transact {
+						origin_kind: OriginKind::SovereignAccount,
```
Comment (vgeddes) [inline parachain/primitives/router/src/lib.rs:107]: If you look at how the `ForeignAssets` pallet is configured in Statemint, you'll notice that it expects the origin kind to be `OriginKind::Xcm`

File: parachain/primitives/router/src/lib.rs [hunk 4]
```diff
@@ -110,3 +178,129 @@ where
 		}
 	}
 }
+
+pub struct EthereumSovereignAccountConvertsFor<AccountId>(PhantomData<AccountId>);
+impl<AccountId: From<[u8; 32]> + Clone> Convert<MultiLocation, AccountId>
```
Comment (bkontur) [inline parachain/primitives/router/src/lib.rs:183]: actual polkadot master does not have `xcm_executor::traits::Convert` but was change to `xcm_executor::traits::ConvertLocation`, so better update deps

File: parachain/primitives/router/src/lib.rs [hunk 5]
```diff
@@ -110,3 +178,129 @@ where
 		}
 	}
 }
+
+pub struct EthereumSovereignAccountConvertsFor<AccountId>(PhantomData<AccountId>);
```
Comment (vgeddes) [inline parachain/primitives/router/src/lib.rs:182]: To be consistent with the naming of `GlobalConsensusParachainConvertsFor`:
```suggestion
pub struct GlobalConsensusEthereumContractConvertsFor<AccountId>(PhantomData<AccountId>);
```

File: parachain/primitives/router/src/lib.rs [hunk 6]
```diff
@@ -110,3 +178,129 @@ where
...
+	fn convert_ref(location: impl Borrow<MultiLocation>) -> Result<AccountId, ()> {
+		log::trace!(
+			target: "xcm::location_conversion",
+			"EthereumSovereignAccountConvertsFor location: {:?}",
+			location.borrow(),
+		);
+
+		if let MultiLocation {
+			parents: 2,
+			interior: X2(GlobalConsensus(Ethereum { chain_id }), AccountKey20 { network, key }),
```
Comment (vgeddes) [inline parachain/primitives/router/src/lib.rs:195]: Should disregard `parents` for GlobalConsensus locations

```suggestion
			interior: X2(GlobalConsensus(Ethereum { chain_id }), AccountKey20 { network, key }), ..
```

File: smoketest/tests/lock_tokens.rs [hunk 7]
```diff
@@ -67,7 +67,16 @@ async fn test_lock_tokens() {
...
 
-    let recipient = MultiLocation::default().encode();
+    // Ferdie 
+    // DE14BzQ1bDXWPKeLoAqdLAm1GpyAWaWF1knF74cEZeomTBM
+    // 0x1cbd2d43530a44705ad088af313e18f80b53ef16b36177cd4b77b846f2a5f07c
+    let recipient = MultiLocation {
+        parents: 0,
+        interior: X1(Junction::AccountId32{
+            network: None,
+            id: [0x1c, 0xbd, 0x2d, 0x43, 0x53, 0x0a, 0x44, 0x70, 0x5a, 0xd0, 0x88, 0xaf, 0x31, 0x3e, 0x18, 0xf8, 0x0b, 0x53, 0xef, 0x16, 0xb3, 0x61, 0x77, 0xcd, 0x4b, 0x77, 0xb8, 0x46, 0xf2, 0xa5, 0xf0, 0x7c],
```
Comment (vgeddes) [inline smoketest/tests/lock_tokens.rs:77]: please rather use `hex_literal:hex!` macro here.

File: parachain/primitives/router/src/lib.rs (hunk 4 above)
Comment (alistair-singh) [reply parachain/primitives/router/src/lib.rs]: Resolved since merge.

File: smoketest/tests/lock_tokens.rs (hunk 7 above)
Comment (alistair-singh) [reply smoketest/tests/lock_tokens.rs]: resolved in [90691c8](https://github.com/Snowfork/snowbridge/pull/846/commits/90691c836b69be19a7c2f0da307d1c4af927344b)

**Review by vgeddes — APPROVED** [review]

**Pull request merged**